[Navigation API] Fix BFCache navigate event timing and activation issues.

basuke opened the pull request:
#### b05914464158fcb242b61edb1c2a547be7a0532c
<pre>
[Navigation API] Fix BFCache navigate event timing and activation issues.
<a href="https://bugs.webkit.org/show_bug.cgi?id=297356">https://bugs.webkit.org/show_bug.cgi?id=297356</a>
<a href="https://rdar.apple.com/158248416">rdar://158248416</a>

Reviewed by Alex Christensen.

This commit addresses two Navigation API issues with Back-Forward Cache operations:

1. Navigation API activation object was not being properly created when pages were restored from
   the back-forward cache, causing:
    - navigation.activation.navigationType was &quot;push&quot; instead of &quot;traverse&quot;.
    - navigation.activation.from was missing/null for same-origin navigations.

2. Navigate events were firing before BFCache page restoration completed, causing:
    event handlers to miss navigate events during traversals to BFCache pages.

The fixes include:
- Add updateForActivation() calls during BFCache restoration in CachedPage::restore() to create
  activation objects with correct traverse navigation type and relationships.
- Defer navigate event dispatch for BFCache traversals until after page restoration.
- Add dispatchPendingNavigateEventForBFCache() to fire events at correct timing.
- Simplify code by using m_pendingNavigationAPIItem as sole state indicator.
- Hide setActivation (renamed from updateForActivation) in updateForReactivation.

The cross-origin security logic ensures navigation.activation.from is null for cross-origin
navigations as expected.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-after-bfcache-cross-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-after-bfcache-cross-origin.html:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-after-bfcache-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-after-bfcache.html:
* Source/WebCore/history/CachedPage.cpp:
(WebCore::CachedPage::restore):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::commitProvisionalLoad):
(WebCore::FrameLoader::continueLoadAfterNavigationPolicy):
(WebCore::FrameLoader::loadItem):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::initializeForNewWindow):
(WebCore::Navigation::setActivation):
(WebCore::Navigation::updateForReactivation):
(WebCore::Navigation::updateForActivation): Renamed to setActivation.
* Source/WebCore/page/Navigation.h:

Canonical link: <a href="https://commits.webkit.org/298765@main">https://commits.webkit.org/298765@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a599208510fa4c3db659d854a49497d6c1bc0bf2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116581 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36242 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26797 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122653 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67153 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c16dd7c6-d0d3-4665-966c-76d9c30a1f0a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118470 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36936 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44830 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88548 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42976 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ac2a3978-68e4-480c-b7dd-656fdc6ab2d2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119530 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29456 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104586 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69008 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28530 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22691 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66320 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98845 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22851 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125787 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43475 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32662 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97219 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43840 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100786 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97011 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24703 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42324 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20251 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39441 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43363 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48954 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42829 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46172 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44534 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2415,6 +2415,18 @@ void FrameLoader::commitProvisionalLoad()
         cachedPage->restore(*page);
 
+        // Dispatch any pending navigate event after BFCache restoration is complete.
+        if (m_pendingNavigationAPIItem) {
```
I think this would be a good place to use std::exchange:
if (RefPtr pendingNavigationItem = std::exchange(m_pendingNavigationItem, nullptr))

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2415,6 +2415,18 @@ void FrameLoader::commitProvisionalLoad()
+            if (RefPtr document = m_frame->document()) {
+                if (RefPtr window = document->window()) {
+                    if (RefPtr navigation = window->navigation(); navigation->frame())
+                        navigation->dispatchTraversalNavigateEvent(Ref { *m_pendingNavigationAPIItem });
```
I don't think we need another ref here if we do my other suggestion.

basuke committed:
[Navigation API] Fix BFCache navigate event timing and activation issues.
https://bugs.webkit.org/show_bug.cgi?id=297356
rdar://158248416

Reviewed by Alex Christensen.

This commit addresses two Navigation API issues with Back-Forward Cache operations:

1. Navigation API activation object was not being properly created when pages were restored from
   the back-forward cache, causing:
    - navigation.activation.navigationType was "push" instead of "traverse".
    - navigation.activation.from was missing/null for same-origin navigations.

2. Navigate events were firing before BFCache page restoration completed, causing:
    event handlers to miss navigate events during traversals to BFCache pages.

The fixes include:
- Add updateForActivation() calls during BFCache restoration in CachedPage::restore() to create
  activation objects with correct traverse navigation type and relationships.
- Defer navigate event dispatch for BFCache traversals until after page restoration.
- Add dispatchPendingNavigateEventForBFCache() to fire events at correct timing.
- Simplify code by using m_pendingNavigationAPIItem as sole state indicator.
- Hide setActivation (renamed from updateForActivation) in updateForReactivation.

The cross-origin security logic ensures navigation.activation.from is null for cross-origin
navigations as expected.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-after-bfcache-cross-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-after-bfcache-cross-origin.html:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-after-bfcache-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-after-bfcache.html:
* Source/WebCore/history/CachedPage.cpp:
(WebCore::CachedPage::restore):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::commitProvisionalLoad):
(WebCore::FrameLoader::continueLoadAfterNavigationPolicy):
(WebCore::FrameLoader::loadItem):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::initializeForNewWindow):
(WebCore::Navigation::setActivation):
(WebCore::Navigation::updateForReactivation):
(WebCore::Navigation::updateForActivation): Renamed to setActivation.
* Source/WebCore/page/Navigation.h:

Canonical link: https://commits.webkit.org/298765@main
